bluemira/structural/model.py

```diff
@@ -0,0 +1,601 @@
+            idn = 6 * node.id_number
+            p_vector[idn : idn + 6] += node.p_vector()
+
+        for element in self.geometry.elements:
+            enf = element.equivalent_node_forces()
+            i = 6 * element.node_1.id_number
+            j = 6 * element.node_2.id_number
+            p_vector[i : i + 6] += enf[:6]
+            p_vector[j : j + 6] += enf[6:]
+        return p_vector
+
+    def _get_reaction_forces(self):
+        """
+        Calculates the reaction forces, applying them to each of the nodes
+        and returning the full vector
+
+        Returns
+        -------
+        reactions: np.array(6*n_nodes)
+            The full reaction vector in global coordinates over all the nodes
+        """
+        reactions = np.zeros(6 * self.geometry.n_nodes)
+
+        for element in self.geometry.elements:
+            p_glob = element.p_vector_glob()
+            element.node_1.reactions += p_glob[:6]
+            element.node_2.reactions += p_glob[6:]
+        for node in self.geometry.nodes:
+            p = node.p_vector()
+            node.reactions += p
+            idn = 6 * node.id_number
+            reactions[idn : idn + 6] = node.reactions
+        return reactions
+
+    def _math_checks(self, k_matrix):
+        """
+        Performs a series of checks on the model boundary conditions and the
+        global stiffness matrix K, prior to the solution of the system of
+        equations.
+
+        Parameters
+        ----------
+        k_matrix: np.array((6*n_nodes, 6*n_nodes))
+            The global stiffness matrix to be checked
+        """
+        if self.n_fixed_dofs < 6:
+            # TODO: check dimensionality of problem (1-D, 2-D, 3-D) and reduce
+            # number of fixed DOFs required accordingly.
+            raise StructuralError(
+                "Insufficient boundary conditions to carry out "
+                f"analysis:\n fixed_dofs: {self.fixed_dofs}"
+                "\n|\tRigid-body motion."
+            )
+        if not self.fixed_dofs.all():
+            raise StructuralError(
+                "Can only solve systems in which all DOFs have "
+                "been constrained at least once."
+            )
+        if not check_matrix_condition(k_matrix, 15):
+            raise StructuralError(
+                "Ill-conditioned or singular stiffness matrix. "
+                "Probably worth checking model boundary "
+                "conditions."
+            )
+
+    def _displacement_check(self, deflections):
+        """
+        Checks to see if the displacements are not too big relative to the
+        size of the model.
+
+        Parameters
+        ----------
+        deflections: np.array(6*n_nodes)
+            The vector of absolute displacements
+        """
+        xmax, xmin, ymax, ymin, zmax, zmin = self.geometry.bounds()
+        dx, dy, dz = xmax - xmin, ymax - ymin, zmax - zmin
+        length = np.sqrt(dx**2 + dy**2 + dz**2)
```

Comment:
np.linalg.norm

Reply:
![image](https://user-images.githubusercontent.com/26097289/161243982-c5848af0-25d7-4aeb-9d6a-a1e4f96730d2.png)
:stuck_out_tongue_winking_eye: 
(for floats!)

Reply:
fair enough!